### ISSUE TITLE
SNOMED server maintenance

### ISSUE THREAD

**Issue by eatyourpeas** [rcpch/rcpch-audit-engine#457]
We currently are using Mark Wardle's excellent Hermes as our SNOMED terminology server and this works well. It has one vulnerability, however, that it is running in a dockerised instance on azure, and ideally the SNOMED CT version needs updating everytime a new one is released from TRUD. This is not easy with the current set up we have and involves manually building a new docker instance with the new version and pushing to azure. 

- [ ] get RCPCH specific TRUD API keys
- [ ] explore hosting an instance of Snowstorm instead
- [ ] set up CRON job or similar to automate the updating of SNOMED CT and notify repository leads
- [ ] refactor the URLS / keys within azure

**Comment by wardle** [comment]
Just spotted this issue. You can, of course, build a pipeline that creates a new docker image every three months and then you switch, but you *can* update in place if you really want to. So, if you're looking for very easy maintenance, you can even run a server and have a cron job that updates the data file at 3 monthly intervals.  

Personally, I have a suite of data files and my application has a top-level configuration file that specifies the version - that means I have clear versioning between application, libraries and supporting data files - and I just bump to the latest release during my application release lifecycle. I release frequently, so it isn't much of an issue, and I generally synchronise the volume containing data files frequently as well so all deployments end up on the same version. I don't run in docker, however, and instead use as a library in-process, so I have a different set-up to you. 

I couldn't use the national services because my requirements mean I'd make far too many requests per second.

Mark

**Comment by wardle** [comment]
I was thinking about your situation and wanting to be able to make easy updates. I have always supported update-in-place, but I realised that during an update process, the data are re-indexed. This could have resulted in processes/threads that were concurrently using those index files to temporarily get inconsistent data. The latest release fixes this, such that it should be possible to update to the latest UK distribution with a single command, and maintain integrity for any server processes that were also using that data file at the same time. You can then run a simple server (AWS/Hetzner etc) and simply update-in-place and avoid some complexity. Hope it helps. 

See [v1.4.1362]( https://github.com/wardle/hermes/releases/tag/v1.4.1362) and (now closed) [issue](https://github.com/wardle/hermes/issues/64)

**Comment by eatyourpeas** [comment]
This is super helpful thank you Mark. We are running a version of SNOMED currently that is a couple of years out of date: we do not need real time versions but I would be happier to be honest that we were serving up to date content. I think we should bump this issue so that I can draw in the large brains of @mbarton @pacharanero @dc2007git and @anchit-chandran

**Comment by wardle** [comment]
Important to note that in my usage, I use a single server - and don't use multiple servers/docker images. I could.. IF I needed to scale, but I don't yet. I keep versioned data files on one server, and then in my deployment configuration I update the version and everything just updates on every deployment of new code. I've just automated the whole pipeline to stop myself making mistakes, and don't try to run multiple servers.

**Comment by wardle** [comment]
And I have used caddy quite nicely to do atomic switches between new services running on different ports. But YMMV.

**Comment by pacharanero** [comment]
Updating this Issue following chats with @eatyourpeas (and coincidentally @wardle) in recent days. cc @mbarton 

We do need to update the SNOMED version. Hermes is deployed in an Azure Container instance (so there is no way we can put Caddy in front of it) however the container instances themselves are kind of hot-swappable I believe. But we might just be able to change the version without having to have a 2-container setup.

I wonder if it could be possible to configure the SNOMED version that is pulled from TRUD with an ENV var, so we set up a replica, add API key and choose new SNOMED version?

What is SNOMED being used for inside E12? How will the new version affect this if - eg there are inactivated concepts in the new version?

**Comment by mbarton** [comment]
> We do need to update the SNOMED version. Hermes is deployed in an Azure Container instance (so there is no way we can put Caddy in front of it) however the container instances themselves are kind of hot-swappable I believe. But we might just be able to change the version without having to have a 2-container setup.

We should be able to create a new one in Azure Container Apps. We can then upgrade by changing the configuration to a new container version and it does a rolling deploy. We use this already for all our apps and also our postcodes.io instance.

> I wonder if it could be possible to configure the SNOMED version that is pulled from TRUD with an ENV var, so we set up a replica, add API key and choose new SNOMED version?

I guess it's too heavy to do this at container start up time? I don't mind a small build process and some GitHub actions to orchestrate it. We don't have to fork any underlying repo, just customise the Docker build (I tried this with Open WebUI and it worked decently https://github.com/rcpch/openwebui-scripts).

> What is SNOMED being used for inside E12? How will the new version affect this if - eg there are inactivated concepts in the new version?

Big list of Epilepsy causes. The data is pulled out of the Hermes instance and stored in the E12 database in a seeding process, so inactive concepts *should* be fine as they should remain in the database even if they're not in the list the next time the seeding runs. I guess we'd need to hide them in the UI somehow but that should probably be under instruction from the E12 audit team themselves rather than automatic.

**Comment by eatyourpeas** [comment]
We should seriously consider using @pacharanero rust project sct (https://github.com/pacharanero/sct)

It is really fast

**Comment by mbarton** [comment]
Yes sorry catching up with the other thread. Can I check my understanding - in E12 we'd replace the SNOMED seed calls to the Hermes server with shell calls to sct? Where would we store the snapshots so CI/CD can access them?

**Comment by pacharanero** [comment]
To be noted as regards using `sct` for this:

* `sct` doesn't yet support SNOMED ECL so if you are using an Epilepsy refset (which I believe you are) it won't be possible to 'auto-expand' an ECL expression.
* I am working on ECL support because it's a prerequisite to being able to build the `sct server` command which would act as a full FHIR Terminology Server.
* `sct` is very new and has no tests and only limited human playtesting. I'd suggest waiting a while for it to mature a little before putting it in a mission-critical project!
* `sct trud` is functionality kindly added by @eatyourpeas and I should be merging this over the weekend. This allows `sct` to pull new SNOMED versions directly from TRUD using the TRUD API. (for CI we would put an API key in GitHub secrets)
* I am working on a `sct codelist` feature which attempts to propose a standard file format for codelists - this allows them to be version controlled, searched for comprehensively across the GitHubs, and shared/remixed much more readily. I might include a way for ECL expansion to be expressed in them as well. Mark Wardle feels that's an essential must-have, and he knows a lot more than I do about this stuff.

To what extent does E12 need a live terminology server? Is SNOMED just used for seeding purposes? I wonder if we could have a call to look at this?

**Comment by wardle** [comment]
Happy to help if I can. Have a look at 'https://github.com/wardle/hermes/actions/runs/24248864588/job/70802443183' - I have several GitHub Action workflows that download, install and index. There's no reason at all why you couldn't build a pipeline to build containers.... my usual approach has not to include data files but run containers with a shared volume and consider the data as the infrastructure side of things.. my own applications now have a weird system in which they just look for local data files and if they're not there they can download them from s3... so I have everything versioned and can recreate everything from app/libraries/data/fixtures etc from any point in time. 

Essentially concepts and terms do get inactivated so we have historical associations to support that - so you can expand ECL and use these to expand to include older concepts - I wrote more about dealing with inactivated concepts here [https://wardle.org/snomed/2021/07/26/making-valuesets.html](https://wardle.org/snomed/2021/07/26/making-valuesets.html). 

My take on opencodelists is documented here: [https://wardle.org/strategy/2023/09/28/codelists.html](https://wardle.org/strategy/2023/09/28/codelists.html).. in reality I think ECL can do all of it, but I quite liked my JSON based DSL for defining this and used it in a project for the MHRA a few years back in which we combined tonnes of data from across the EU, so using SNOMED as the ground truth from lots of crazy systems was invaluable.  The JSON was more readable for the analysts than ECL. 

Personally, if you're not using a terminology server live, I'd build a CI/CD pipeline that downloads SNOMED, does whatever you need to get the data and then build your code lists and save them somewhere for later consumption e.g. in S3. Reproducible and versioned and dated. But you do need historical associations if you're going to deal with real-life data as SNOMED deletes the relationships from inactivated concepts, so any inference whether simple or complex will fail with those concepts.

**Comment by wardle** [comment]
Search "history supplements"!

**Comment by wardle** [comment]
To be clear the reason I needed more than ECL was ATC codes which were used across many other centres, so https://github.com/wardle/codelists consumed SNOMED CT and the UK dm+d - latter has ATC maps that have not been converted into a SNOMED cross map. If I'd been more thoughtful I could have written a tool to generate those cross maps in RF2 format instead and just used Hermes but easy to say in retrospect! As it is I shoved dm+d into a SQLite database and consumed it - ie composition of simpler tools.
